wgpu-core/src/command/query.rs

```diff
@@ -214,133 +213,110 @@ impl<A: HalApi> QuerySet<A> {
-            query_set_id,
-            SimplifiedQueryType::PipelineStatistics,
-            query_index,
-            reset_state,
-        )?;
-
-        if let Some((_old_id, old_idx)) = active_query.replace((query_set_id, query_index)) {
-            return Err(QueryUseError::AlreadyStarted {
-                active_query_index: old_idx,
-                new_query_index: query_index,
-            });
-        }
-
-        unsafe {
-            // If we don't have a reset state tracker which can defer resets, we must reset now.
-            if needs_reset {
-                raw_encoder.reset_queries(self.raw(), query_index..(query_index + 1));
-            }
-            raw_encoder.begin_query(query_set, query_index);
+    unsafe {
+        // If we don't have a reset state tracker which can defer resets, we must reset now.
+        if needs_reset {
+            raw_encoder.reset_queries(query_set.raw(), query_index..(query_index + 1));
         }
-
-        Ok(())
+        raw_encoder.begin_query(query_set.raw(), query_index);
     }
+
+    Ok(())
 }
 
 pub(super) fn end_occlusion_query<A: HalApi>(
     raw_encoder: &mut A::CommandEncoder,
-    storage: &Storage<QuerySet<A>>,
-    active_query: &mut Option<(id::QuerySetId, u32)>,
+    active_query: &mut Option<(Arc<QuerySet<A>>, u32)>,
 ) -> Result<(), QueryUseError> {
-    if let Some((query_set_id, query_index)) = active_query.take() {
-        // We can unwrap here as the validity was validated when the active query was set
-        let query_set = storage.get(query_set_id).unwrap();
-
+    if let Some((query_set, query_index)) = active_query.take() {
         unsafe { raw_encoder.end_query(query_set.raw.as_ref().unwrap(), query_index) };
-
         Ok(())
     } else {
         Err(QueryUseError::AlreadyStopped)
     }
 }
 
-pub(super) fn end_pipeline_statistics_query<A: HalApi>(
+pub(super) fn validate_and_begin_pipeline_statistics_query<A: HalApi>(
+    query_set: Arc<QuerySet<A>>,
     raw_encoder: &mut A::CommandEncoder,
-    storage: &Storage<QuerySet<A>>,
-    active_query: &mut Option<(id::QuerySetId, u32)>,
+    query_index: u32,
+    reset_state: Option<&mut QueryResetMap<A>>,
+    active_query: &mut Option<(Arc<QuerySet<A>>, u32)>,
 ) -> Result<(), QueryUseError> {
-    if let Some((query_set_id, query_index)) = active_query.take() {
-        // We can unwrap here as the validity was validated when the active query was set
-        let query_set = storage.get(query_set_id).unwrap();
+    let needs_reset = reset_state.is_none();
+    query_set.validate_query(
+        SimplifiedQueryType::PipelineStatistics,
+        query_index,
+        reset_state,
+    )?;
+
+    if let Some((_old, old_idx)) = active_query.take() {
+        return Err(QueryUseError::AlreadyStarted {
+            active_query_index: old_idx,
+            new_query_index: query_index,
+        });
+    }
+    let query_set = &active_query.insert((query_set, query_index)).0;
```

Comment:
nitpick: It's easier to follow this code when you mirror the tuple structure b/w LHS and RHS:

```suggestion
    let (query_set, _) = &active_query.insert((query_set, query_index));
```

Pushed a `fixup!` for this and the case in `validate_and_begin_occlusion_query` with 96432ec4d.

Reply:
good style advice, should do that more often like so